iwyu_ast_util.cc

```diff
@@ -951,13 +953,57 @@ bool IsDefaultNewOrDelete(const FunctionDecl* decl,
+  // void* operator new[]( std::size_t count );
+  // void* operator new  ( std::size_t count, std::align_val_t al ); (since C++17)
+  // void* operator new[]( std::size_t count, std::align_val_t al ); (since C++17)
+  //
+  // Likewise, the following variants of operator delete[2] are implicitly
+  // defined in every translation unit and should not require including <new>.
+  //
+  // void operator delete  ( void* ptr ) throw(); (until C++11)
+  // void operator delete  ( void* ptr ) noexcept; (since C++11)
+  // void operator delete[]( void* ptr ) throw(); (until C++11)
+  // void operator delete[]( void* ptr ) noexcept; (since C++11)
+  // void operator delete  ( void* ptr, std::align_val_t al ) noexcept; (since C++17)
+  // void operator delete[]( void* ptr, std::align_val_t al ) noexcept; (since C++17)
+  // void operator delete  ( void* ptr, std::size_t sz ) noexcept; (since C++14)
+  // void operator delete[]( void* ptr, std::size_t sz ) noexcept; (since C++14)
+  // void operator delete  ( void* ptr, std::size_t sz,
+  //                         std::align_val_t al ) noexcept; (since C++17)
+  // void operator delete[]( void* ptr, std::size_t sz,
+  //                         std::align_val_t al ) noexcept; (since C++17)
+  // void operator delete  ( void* ptr, const std::nothrow_t& tag ) throw(); (until C++11)
+  // void operator delete  ( void* ptr, const std::nothrow_t& tag ) noexcept; (since C++11)
+  // void operator delete[]( void* ptr, const std::nothrow_t& tag ) throw(); (until C++11)
+  // void operator delete[]( void* ptr, const std::nothrow_t& tag ) noexcept; (since C++11)
+  //
+  // FunctionDecl::isReplaceableGlobalAllocationFunction() does a great job of
+  // parsing the variants of operator new/delete and returning true if a
+  // particular variant is "replaceable". It just so happens that all implicitly
+  // defined variants are also replaceable. However, there are four variants of
+  // operator new that are replaceable but not implicitly defined; we must look
+  // for these explicitly and filter them out.
+  //
+  // 1. https://en.cppreference.com/w/cpp/memory/new/operator_new
+  // 2. https://en.cppreference.com/w/cpp/memory/new/operator_delete
+  Optional<unsigned> alignmentParam;
+  bool isNoThrow = false;
+  if (!decl->isReplaceableGlobalAllocationFunction(&alignmentParam, &isNoThrow))
     return false;
 
-  return true;
+  // These variants are replaceable but not implicitly defined.
+  //
+  // void* operator new  ( std::size_t count, const std::nothrow_t& tag );
+  // void* operator new[]( std::size_t count, const std::nothrow_t& tag );
+  // void* operator new  ( std::size_t count,
+  //                       std::align_val_t al, const std::nothrow_t& ); (since C++17)
+  // void* operator new[]( std::size_t count,
+  //                       std::align_val_t al, const std::nothrow_t& ); (since C++17)
+  return StartsWith(decl_name, "operator delete") || !isNoThrow;
```

Comment:
I see `isReplaceable...` uses `(getDeclName().getNameKind() != DeclarationName::CXXOperatorName)` and `getDeclName().getCXXOverloadedOperator()` to use enums instead of string comparisons. That does seem a little nicer, can you see if that's applicable here too, outside the function?

Reply:
Yeah I can do that, though I don't think we need the `getDeclName().getNameKind()` check, as having returned true from `isReplaceable...`, we're guaranteed that this is a `CXXOperatorName`.